### PR TITLE
Refactor/avoid recursion

### PR DESCRIPTION
**Pull request by perweij**
The recursive nature of `-partition-after-pred` causes problems when processing long lists where these limits are exceeded:
 `max-lisp-eval-depth`
 `max-specpdl-size`

I have tried to rewrite it as an iterative macro instead. It seems to work, and passes the tests, but if you have time, please check the code for possible improvements.

### CODE REVIEW

File: dash.el [hunk 1]
```diff
@@ -1170,20 +1170,28 @@ returns the header value, but only after seeing at least one
 other value (the body)."
   (--partition-by-header (funcall fn it) list))
 
-(defun -partition-after-pred (pred list)
+(defmacro -partition-after-pred (pred list)
```
Comment (basil-conto) [inline dash.el:1173]: It's not possible to turn a function into a macro with the same name, as that will break its existing callers. Also, Dash follows the naming convention where functions start with a single dash (e.g. `-foo`) and their macro counterparts start with two dashes (e.g. `--foo`).

I therefore suggest that, if you deem a macro is necessary, you provide that separately, and optionally reimplement the function in terms of the macro.

File: dash.el [hunk 2]
```diff
@@ -1170,20 +1170,28 @@ returns the header value, but only after seeing at least one
...
-      (if (funcall pred (car list))
-          ;;split after (car list)
-          (cons (list (car list))
-                rest)
-
-        ;;don't split after (car list)
-        (cons (cons (car list)
-                    (car rest))
-              (cdr rest))))))
+  (declare (debug (pred pred)))
```
Comment (basil-conto) [inline dash.el:1175]: AFAIK `pred` isn't a valid Edebug spec; see [`(info "(elisp) Instrumenting Macro Calls")`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Instrumenting-Macro-Calls.html) for info on `debug` declarations.

File: dash.el [hunk 3]
```diff
@@ -1170,20 +1170,28 @@ returns the header value, but only after seeing at least one
...
+  (let ((r (make-symbol "result"))
+        (s (make-symbol "sublist"))
+        (l (make-symbol "list")))
+    `(let ((,l ,list))
+       (when ,l
+         (let* ((,r nil)
+		(,s nil))
```
Comment (basil-conto) [inline dash.el:1182]: Please consistently indent with spaces, in accordance with the `indent-tabs-mode` setting in the project's `.dir-locals.el` file.

**Review by basil-conto — CHANGES_REQUESTED** [review]
Thanks, it's always good to avoid recursion in Elisp if possible! :) But this still needs a bit of work. ;)

This probably counts as a non-trivial change (in terms of lines of code), so in order to install it we'd need you to assign the copyright to the Free Software Foundation, as Dash is copyrighted by the FSF.  Do you already have such an assignment on file?  If not, would you be willing to submit one?  The process is relatively simple and painless, and once it's done you'll be able to contribute to Emacs itself and other core packages as well. See [`(info "(emacs) Copyright Assignment")`](https://www.gnu.org/software/emacs/manual/html_node/emacs/Copyright-Assignment.html) for more info on this.  You can find the relevant form here: https://git.savannah.gnu.org/cgit/gnulib.git/tree/doc/Copyright/request-assign.future.

TIA.

File: dash.el (hunk 1 above)
Comment (Fuco1) [reply dash.el:1173]: The macro version should then be anaphoric in the predicate to follow the rest.

Personally these days I prefer to write the macros in terms of a function, the performance impact seems ignorable (until someone finds it slow enough).

Comment (basil-conto) [reply dash.el:1173]: > (until someone finds it slow enough)

;)

**Comment by perweij** [comment]
Hi! Many thanks for checking this so quickly!

> If not, would you be willing to submit one? 

Done - I'll update here when I have received a reply.

File: dash.el (hunk 1 above)
Comment (perweij) [reply dash.el:1173]: Hi! Thank you very much for the review! I'll rework it accordingly and be back,

File: dash.el (hunk 2 above)
Comment (perweij) [reply dash.el:1175]: Good catch.

Comment (perweij) [reply dash.el]: I removed it now, so should be OK.

File: dash.el (hunk 1 above)
Comment (perweij) [reply dash.el]: Now it is a function again, and a new anaphoric macro has been added:
  `(defun -partition-after-pred (pred list) ...)`
  `(defmacro --partition-after-pred (form list) ...)`

File: dash.el (hunk 3 above)
Comment (perweij) [reply dash.el]: Fixed, good catch.

**Comment by perweij** [comment]
sorry @basil-conto I did not just deliberately request a new review - misunderstood the refresh icon, thought it would reload stuff but instead it sent you a request. Again, sorry.

**Comment by basil-conto** [comment]
No worries! Now I know what it does too :)

**Comment by basil-conto** [comment]
> Done - I'll update here when I have received a reply.

What's the status on this @perweij?  If you're waiting more than a week or two for a response from the copyright clerk, feel free to ping them.

**Comment by basil-conto** [comment]
Ping.

**Comment by perweij** [comment]
Hi! Now I have assigned the copyright to the Free Software Foundation. The process is completed, and I have received the final PDF file from FSF, so we should be able to proceed.  Sorry for the looong delay!

**Review by basil-conto — APPROVED** [review]

**Comment by basil-conto** [comment]
> Sorry for the looong delay!

No worries, and thanks for the nice and efficient implementation!

**Pull request merged**